File: src/main/java/com/example/travelator/money/java/Money.java
package com.example.travelator.money.java;

import java.math.BigDecimal;
import java.util.Currency;
import java.util.Objects;

public class Money {
	private final BigDecimal amount;
	private final Currency currency;
	
	private Money(BigDecimal amount, Currency currency) {
		this.amount = amount;
		this.currency = currency;
	}
	
	public static Money of(BigDecimal amount, Currency currency) {
		return new Money(amount.setScale(currency.getDefaultFractionDigits()), currency);
	}
	
	public Money add(Money other) {
		if (!this.currency.equals(other.currency)) {
			throw new IllegalArgumentException("cannot add Money values in different currencies");
		}
		return new Money(amount.add(other.amount), currency);
	}
	
	public BigDecimal getAmount() {
		return amount;
	}
	
	public Currency getCurrency() {
		return currency;
	}
	
	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (!(o instanceof Money money)) return false;
		return Objects.equals(amount, money.amount)
			&& Objects.equals(currency, money.currency);
	}
	
	@Override
	public int hashCode() {
		return Objects.hash(amount, currency);
	}
	
	@Override
	public String toString() {
		return amount.toString() + " " + currency.getCurrencyCode();
	}
}

File: src/main/java/com/example/travelator/money/kotlin/Money.kt
package com.example.travelator.money.kotlin

import java.math.BigDecimal
import java.util.*

class Money
/**
 * 주 생성자를 사용하지만 생성자 앞에 private를 추가
 */
private constructor(
    val amount: BigDecimal,
    val currency: Currency
) {
    override fun equals(other: Any?): Boolean {
        if(this === other) return true
        if(other == null || javaClass != other.javaClass) return false
        val money = other as Money
        return amount == money.amount && currency == money.currency
    }

    override fun hashCode(): Int {
        return Objects.hash(amount, currency)
    }

    override fun toString(): String {
        return amount.toString() + " " + currency.currencyCode
    }

    fun add(other: Money): Money {
        require(currency == other.currency) { "cannot add Money values of different currencies" }
        return Money(amount.add(other.amount), currency)
    }

    companion object {
        @JvmStatic
        fun of(amount: BigDecimal, currency: Currency): Money {
            return Money(amount.setScale(currency.defaultFractionDigits), currency)
        }
    }

}